File: api/src/main/java/dev/akif/crud/simpler/SimplerController.java
package dev.akif.crud.simpler;

import dev.akif.crud.CRUDController;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.io.Serializable;

/**
 * Simpler variant of {@link CRUDController} where create/update models and DTOs are just models
 * <p>
 * This is meant to be extended from a @{@link RestController} class,
 * ideally also with a @{@link RequestMapping} with some path prefix for the endpoints.
 *
 * @param <I>      Id type of the data
 * @param <E>      Entity type of the data which is a {@link SimplerEntity}
 * @param <M>      Model type of the data which is a {@link SimplerModel}
 * @param <Mapper> Mapper type of the data which is a {@link SimplerMapper}
 * @param <S>      Service type of the data which is a {@link SimplerService}
 */
public abstract class SimplerController<
        I extends Serializable,
        E extends SimplerEntity<I, E>,
        M extends SimplerModel<I, E, M>,
        Mapper extends SimplerMapper<I, E, M>,
        S extends SimplerService<I, E, M, Mapper>> extends CRUDController<I, E, M, M, M, M, M, M, Mapper, Mapper, S> {
    /**
     * Constructor to provide type name and dependencies to this controller
     *
     * @param type    Type name of the data this controller manages
     * @param service Service dependency of this controller
     * @param mapper  Mapper dependency of this controller
     */
    protected SimplerController(final String type, final S service, final Mapper mapper) {
        super(type, service, mapper);
    }
}

File: api/src/main/kotlin/dev/akif/crud/simpler/SimplerController.kt
package dev.akif.crud.simpler

import dev.akif.crud.CRUDController
import java.io.Serializable

/**
 * Simpler variant of [CRUDController] where create/update models and DTOs are just models
 *
 * This is meant to be extended from a **@RestController** class,
 * ideally also with a **@RequestMapping** with some path prefix for the endpoints.
 *
 * @param I      Id type of the data
 * @param E      Entity type of the data which is a [SimplerEntity]
 * @param M      Model type of the data which is a [SimplerModel]
 * @param Mapper Mapper type of the data which is a [SimplerMapper]
 * @param R      Repository type of the data which is a [SimplerRepository]
 * @param S      Service type of the data which is a [SimplerService]
 */
abstract class SimplerController<
        I : Serializable,
        E : SimplerEntity<I>,
        M : SimplerModel<I>,
        out Mapper : SimplerMapper<I, E, M>,
        out R : SimplerRepository<I, E>,
        out S : SimplerService<I, E, M, R, Mapper>>(
    override val typeName: String,
    override val service: S,
    override val mapper: Mapper
) : CRUDController<I, E, M, M, M, M, M, M, Mapper, Mapper, R, S>(typeName, service, mapper)
